File: mobile/lib/domain/pricing/kalkulator_tarif.dart
import 'package:flutter/foundation.dart';

import '../enums.dart';
import '../models/tarif.dart';

/// Satu baris rincian pembentuk harga.
///
/// Harga Jalur A ditampilkan terurai, bukan sebagai satu angka gelap. Klien
/// yang bisa melihat "tarif dasar sekian, jarak sekian" tidak perlu bertanya
/// ke admin, dan itulah gunanya kalkulator harga otomatis.
@immutable
class RincianTarif {
  const RincianTarif({required this.label, required this.nominal});

  final String label;
  final int nominal;
}

@immutable
class HasilTarif {
  const HasilTarif({required this.rincian, required this.total});

  final List<RincianTarif> rincian;
  final int total;
}

/// Kalkulator harga Jalur A.
///
/// Fungsi murni tanpa ketergantungan ke Flutter maupun jaringan, supaya bisa
/// diuji langsung dan dipindahkan ke backend apa adanya nanti. Kembaran
/// `KalkulatorTarif.Hitung` di server, dan bentuknya sengaja sama supaya
/// perbedaan hasil antara keduanya ketahuan sebagai perbedaan angka, bukan
/// sebagai perbedaan cara memanggil.
///
/// [Tarif] diterima sebagai parameter wajib di setiap fungsi, bukan dibaca
/// dari konstanta statis: angkanya sekarang datang dari server dan bisa
/// diubah admin. Tidak ada nilai bawaan untuk parameter ini dengan sengaja —
/// layar yang lupa mengambil tarif dari [tarifProvider] akan gagal saat
/// dikompilasi, bukan diam-diam menghitung dengan angka yang salah.
class KalkulatorTarif {
  const KalkulatorTarif._();

  /// Harga satu layanan Jalur A.
  ///
  /// Yang mengikat tetap hitungan server. Hitungan di sini gunanya menampilkan
  /// rincian sebelum klien memesan, supaya ia tidak perlu menekan tombol dulu
  /// untuk tahu berapa yang akan ditagih.
  static HasilTarif hitung(ServiceType serviceType, double? jarakKm, Tarif tarif) {
    if (serviceType.track != OrderTrack.jalurA) {
      throw StateError(
        '${serviceType.name} adalah Jalur B, harganya ditentukan admin lewat '
        'penawaran',
      );
    }

    return switch (serviceType) {
      ServiceType.anterJemput => anterJemput(
        jarakKm: _wajibJarak(serviceType, jarakKm),
        tarif: tarif,
      ),
      ServiceType.jastipBarang => jastipBarang(
        jarakKm: _wajibJarak(serviceType, jarakKm),
        tarif: tarif,
      ),
      ServiceType.jastipMakanan => jastipMakanan(tarif: tarif),
      _ => throw StateError('${serviceType.name} belum punya rumus tarif'),
    };
  }

  static double _wajibJarak(ServiceType serviceType, double? jarakKm) {
    if (jarakKm == null) {
      throw StateError('${serviceType.name} butuh jarak untuk dihitung');
    }
    if (jarakKm.isNaN || jarakKm.isInfinite) {
      throw StateError('Jarak bukan angka yang sah');
    }
    return jarakKm;
  }

  /// Jastip Makanan: fee tetap, harga makanannya dibayar terpisah.
  ///
  /// Tidak bergantung jarak, jadi tidak menerima jarak sama sekali. Parameter yang
  /// diterima lalu diabaikan akan dibaca orang berikutnya sebagai sesuatu yang
  /// berpengaruh.
  static HasilTarif jastipMakanan({required Tarif tarif}) => HasilTarif(
    rincian: [
      RincianTarif(label: 'Ongkos jasa titip', nominal: tarif.jastipMakananFee),
    ],
    total: tarif.jastipMakananFee,
  );

  /// Anter jemput: tarif dasar + ongkos jarak.
  ///
  /// [jarakKm] diisi sendiri oleh klien sebagai perkiraan. Ini konsekuensi dari
  /// keputusan memakai alamat teks bebas, bukan pin peta (rencana capstone
  /// bagian 14.8), tanpa peta, sistem tidak punya cara menghitung jarak
  /// sendiri. Selisih kecil diselesaikan runner dan klien di lapangan.
  static HasilTarif anterJemput({required double jarakKm, required Tarif tarif}) {
    final jarakDipakai = jarakKm.clamp(
      tarif.anjemJarakMinimalKm,
      tarif.anjemJarakMaksimalKm,
    );
    final ongkosJarak = (jarakDipakai * tarif.anjemTarifPerKm).round();

    return HasilTarif(
      rincian: [
        RincianTarif(label: 'Tarif dasar', nominal: tarif.anjemTarifDasar),
        RincianTarif(
          label: 'Jarak ${_formatJarak(jarakDipakai)} km',
          nominal: ongkosJarak,
        ),
      ],
      total: tarif.anjemTarifDasar + ongkosJarak,
    );
  }

  /// Jastip Barang: ongkos jasa titip + ongkos jarak.
  ///
  /// Harga barangnya sendiri TIDAK dihitung di sini. Berapa harga barang baru
  /// diketahui setelah runner sampai di tempat, sementara sistem menuntut
  /// pembayaran di depan. Bentrokan itu belum diputuskan mitra (rencana
  /// capstone bagian 14.7a), jadi yang ditagih aplikasi untuk sekarang hanya
  /// jasanya. Begitu mitra menjawab, rumus ini yang berubah, bukan layarnya.
  ///
  /// Batas jarak untuk sementara memakai batas anter jemput, karena mitra
  /// belum memberi angka sendiri untuk jastip.
  static HasilTarif jastipBarang({required double jarakKm, required Tarif tarif}) {
    final jarakDipakai = jarakKm.clamp(
      tarif.anjemJarakMinimalKm,
      tarif.anjemJarakMaksimalKm,
    );
    final ongkosJarak = (jarakDipakai * tarif.jastipBarangTarifPerKm).round();

    return HasilTarif(
      rincian: [
        RincianTarif(label: 'Ongkos jasa titip', nominal: tarif.jastipBarangFee),
        RincianTarif(
          label: 'Jarak ${_formatJarak(jarakDipakai)} km',
          nominal: ongkosJarak,
        ),
      ],
      total: tarif.jastipBarangFee + ongkosJarak,
    );
  }

  static String _formatJarak(double jarak) =>
      jarak == jarak.roundToDouble()
      ? jarak.toStringAsFixed(0)
      : jarak.toStringAsFixed(1).replaceAll('.', ',');
}

File: backend/src/UpnvjSuruh.Api/Pricing/KalkulatorTarif.cs
using UpnvjSuruh.Api.Domain;

namespace UpnvjSuruh.Api.Pricing;

/// <summary>
/// Menghitung harga Jalur A di server.
///
/// Aplikasi punya kalkulator yang sama supaya klien melihat rinciannya sebelum memesan,
/// tapi angka itu tidak pernah dipercaya. Yang mengikat adalah hasil hitungan di sini.
/// Server yang menerima harga jadi dari badan permintaan berarti siapa pun bisa memesan
/// seharga satu rupiah, dan yang seperti itu tidak ketahuan sampai uangnya dihitung.
///
/// Yang diterima dari klien cuma jaraknya, dan itu pun dijepit ke rentang yang wajar.
/// Jarak diisi sendiri karena alamatnya teks bebas, bukan pin peta (bagian 14.8), jadi
/// tanpa peta sistem memang tidak punya cara menghitungnya sendiri.
///
/// <see cref="TarifSetting"/> diterima sebagai parameter, bukan dibaca dari konstanta
/// statis: angkanya sekarang bisa diubah admin lewat dashboard, dan kalkulator ini tetap
/// murni dan gampang diuji kalau tidak perlu tahu dari mana angkanya datang. Tidak ada
/// nilai bawaan untuk parameter ini dengan sengaja — pemanggil yang lupa memuat tarif
/// sekarang dari basis data akan gagal saat kompilasi, bukan diam-diam menghitung dengan
/// angka yang salah.
/// </summary>
public class KalkulatorTarif
{
    public HasilTarif Hitung(ServiceType serviceType, double? jarakKm, TarifSetting tarif)
    {
        if (serviceType.Track() != OrderTrack.JalurA)
        {
            throw new ArgumentException(
                $"{serviceType} adalah Jalur B, harganya ditentukan admin lewat penawaran.",
                nameof(serviceType));
        }

        return serviceType switch
        {
            ServiceType.AnterJemput => AnterJemput(WajibJarak(serviceType, jarakKm), tarif),
            ServiceType.JastipBarang => JastipBarang(WajibJarak(serviceType, jarakKm), tarif),
            ServiceType.JastipMakanan => JastipMakanan(tarif),
            _ => throw new ArgumentOutOfRangeException(nameof(serviceType), serviceType, null),
        };
    }

    private static double WajibJarak(ServiceType serviceType, double? jarakKm)
    {
        if (jarakKm is null)
        {
            throw new ArgumentException($"{serviceType} butuh jarak untuk dihitung.", nameof(jarakKm));
        }

        if (double.IsNaN(jarakKm.Value) || double.IsInfinity(jarakKm.Value))
        {
            throw new ArgumentException("Jarak bukan angka yang sah.", nameof(jarakKm));
        }

        return jarakKm.Value;
    }

    private static HasilTarif AnterJemput(double jarakDiminta, TarifSetting tarif)
    {
        // Dijepit, bukan ditolak. Jarak di bawah minimal tetap order yang wajar, dan jarak
        // di atas maksimal sudah di luar wilayah layanan sehingga ditagih pada batasnya
        // lalu diselesaikan admin. Yang penting angka apa pun yang dikirim tidak bisa
        // membuat harganya nol atau meledak.
        var jarakDipakai = Math.Clamp(jarakDiminta, tarif.AnjemJarakMinimalKm, tarif.AnjemJarakMaksimalKm);
        var ongkosJarak = Math.Round((decimal)jarakDipakai * tarif.AnjemTarifPerKm, MidpointRounding.AwayFromZero);

        return new HasilTarif(
            [
                new RincianTarif("Tarif dasar", tarif.AnjemTarifDasar),
                new RincianTarif($"Jarak {FormatJarak(jarakDipakai)} km", ongkosJarak),
            ],
            tarif.AnjemTarifDasar + ongkosJarak);
    }

    private static HasilTarif JastipBarang(double jarakDiminta, TarifSetting tarif)
    {
        // Batas jarak jastip barang sengaja memakai batas anter jemput yang sama, mengikuti
        // KalkulatorTarif.jastipBarang di aplikasi mobile: mitra belum memberi angka batas
        // sendiri untuk jastip (rencana bagian 14.7a).
        var jarakDipakai = Math.Clamp(jarakDiminta, tarif.AnjemJarakMinimalKm, tarif.AnjemJarakMaksimalKm);
        var ongkosJarak = Math.Round((decimal)jarakDipakai * tarif.JastipBarangTarifPerKm, MidpointRounding.AwayFromZero);

        return new HasilTarif(
            [
                new RincianTarif("Ongkos jasa titip", tarif.JastipBarangFee),
                new RincianTarif($"Jarak {FormatJarak(jarakDipakai)} km", ongkosJarak),
            ],
            tarif.JastipBarangFee + ongkosJarak);
    }

    private static HasilTarif JastipMakanan(TarifSetting tarif) =>
        new([new RincianTarif("Ongkos jasa titip", tarif.JastipMakananFee)], tarif.JastipMakananFee);

    private static string FormatJarak(double jarak) =>
        jarak == Math.Round(jarak)
            ? jarak.ToString("0")
            : jarak.ToString("0.0").Replace('.', ',');
}
